rpc/daemon.go

```diff
@@ -0,0 +1,70 @@
+package rpc
+
+import (
+	"net/http"
+	"runtime/debug"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/athanorlabs/atomic-swap/cliutil"
+	"github.com/athanorlabs/atomic-swap/common"
+)
+
+// DaemonService handles general daemon operations
+type DaemonService struct {
+	server *Server
+	cfg    *Config
+}
+
+// NewDaemonService ...
+func NewDaemonService(server *Server, cfg *Config) *DaemonService {
+	return &DaemonService{
+		server,
+		cfg,
+	}
+}
+
+// ShutdownRequest ...
+type ShutdownRequest struct{}
+
+// ShutdownResponse ...
+type ShutdownResponse struct{}
+
+// Shutdown swapd
+func (s *DaemonService) Shutdown(_ *http.Request, req *ShutdownRequest, resp *ShutdownResponse) error {
+	return s.server.Stop()
+}
+
+// VersionRequest ...
+type VersionRequest struct{}
+
+// VersionResponse ...
+type VersionResponse struct {
+	SwapdVersion    string             `json:"swapdVersion" validate:"required"`
+	P2PVersion      string             `json:"p2pVersion" validate:"required"`
+	Env             common.Environment `json:"env" validate:"required"`
+	SwapCreatorAddr ethcommon.Address  `json:"swapCreatorAddress" validate:"required"`
+}
+
+// Version returns version & misc info about swapd and its dependencies
+func (s *DaemonService) Version(_ *http.Request, req *VersionRequest, resp *VersionResponse) error {
+	resp.SwapdVersion = cliutil.GetVersion()
+
+	resp.P2PVersion = func() string {
```

Comment:
For any given versioned build of `swapd`, the version of the P2P library is fixed, so we don't need that particular piece of information.

What we want is the `protocolID` passed to the p2p network layer (it is passed in `RunSwapDaemon` to the `net.NewHost` call). This identifier determines whether 2 versions of `swapd` can communicate with each other for running swaps. Two different versions of `swapd` can still communicate and run swaps together if we haven't bumped that protocol version. @noot Correct me if I'm wrong on anything in this paragraph. I was surprised when you told me that bootnodes can provide DHT support for multiple protocol IDs simultaneously.